### PR TITLE
Adding US State dropdown to account creation to support CPA

### PR DESCRIPTION
**Pull request by daynew**
For the CPA project, we need to know if the student creating an account lives in Colorado. To do this, we need to ask the student what US state they live in if we detect they are using a computer in the US.

* Add support for a "us_state" and "country_code" serializable attribute in the User model.
* Add a “State” input for the[ section account creation page](http://studio.code.org/join) if the student is in the US.
* Add a “State” input for the [main account creation page](https://studio.code.org/users/sign_up) if the student is in the US.

### `/join`
Join a section screen showing the new US State dropdown
### `/join` error
Join a section screen showing error in the inputs which have a yellow background and red outline

### `/join` before & after spacing fixes
Note the inconsistent spacing after the "Age" input.

#### before
Join a section screen showing bad spacing after the Age input

#### after
Join a section screen showing good spacing after the Age input

### `/users/sign_up`
Account creation screen showing the new US State dropdown

## Links
* [JIRA Task](https://codedotorg.atlassian.net/browse/P20-137)

## Testing story
* Some unit tests to test the User object will accept the new attributes
* Manually tested on localhost
  * http://localhost-studio.code.org:3000/join
  * http://localhost-studio.code.org:3000/users/sign_up

## Deployment strategy
* The UI elements are hidden behind the `cpa_experience` feature flag. The User model changes should only be activated if form inputs hidden behind the feature flag are present.
## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/assets/stylesheets/scaffolds.scss
Comment (elf-code) [inline dashboard/app/assets/stylesheets/scaffolds.scss:36]: does this change affect other inputs? Also can you include a screenshot of it?

File: dashboard/app/models/user.rb
Comment (elf-code) [inline dashboard/app/models/user.rb:2615]: weird corner case - what about US Territories? Guam, the Virgin Islands, American Samoa, etc.

File: dashboard/app/views/followers/student_user_new.html.haml
Comment (elf-code) [inline dashboard/app/views/followers/student_user_new.html.haml:82]: can you add a before/after screenshot to make sure this <br> blends in?

**Review by elf-code — COMMENTED** [review]
Looks like you've got some merge conflicts. The code looks good though. I'll hold off on accept until the merge conflicts are resolved.

File: dashboard/app/views/devise/registrations/edit.html.haml
Comment (carl-codeorg) [inline dashboard/app/views/devise/registrations/edit.html.haml]: Does this also need to be gated behind cpa_experience?

File: dashboard/app/models/user.rb
Comment (carl-codeorg) [inline dashboard/app/models/user.rb:2615]: Should this const be moved to shared_constants.rb or maybe the application_helper? It's kind of long, and I could see it being used in the future for other address-related things.

**Comment by carl-codeorg** [comment]
> Looks like you've got some merge conflicts. The code looks good though. I'll hold off on accept until the merge conflicts are resolved.

Merge conflicts are my fault - I merged the gender stuff late yesterday and it touches a bunch of the same files

File: dashboard/app/views/devise/registrations/edit.html.haml
Comment (daynew) [reply dashboard/app/views/devise/registrations/edit.html.haml]: Oops this actually shouldn't be in here. This is something I made when trying to debug the issues we were having with updating serialized attributes! I will remove this.

File: dashboard/app/models/user.rb
Comment (daynew) [reply dashboard/app/models/user.rb:2615]: I chatted with product and legal, we are okay with just the 50 states and Washington DC. I also ran some quick experiments with our geocoder, and it looks like it identifies those territories as their own countries. Students in those territories shouldn't see this UI because we only show it IP's in the United States IP range.

File: dashboard/app/assets/stylesheets/scaffolds.scss
Comment (daynew) [reply dashboard/app/assets/stylesheets/scaffolds.scss:36]: Screenshots updated. Hard to know all possible inputs this could be affecting, but it should be a good change because it makes it consistent with the /users/signup styling. Also it's a good a11y pattern to add a border around inputs which are in error.

File: dashboard/app/views/followers/student_user_new.html.haml
Comment (daynew) [reply dashboard/app/views/followers/student_user_new.html.haml:82]: Done. I understand your concern, it looks like I'm adding rogue breaks. But the styling on this page was actually broken before I made any changes to it. See the screenshots.

File: dashboard/app/models/user.rb
Comment (daynew) [reply dashboard/app/models/user.rb:2615]: I think this is specific to the `us_state` column in the User model. A more generic solution would probably want to support territories and other things as well.

Comment (carl-codeorg) [reply dashboard/app/models/user.rb:2615]: Makes sense. Do you know why we treat `full_address` as a string instead of doing a more complete address object, with properties like street address, country, state, zip, etc?

Comment (daynew) [reply dashboard/app/models/user.rb:2615]: nope :(

**Comment by daynew** [comment]
Resolved merge conflicts :)

**Review by carl-codeorg — APPROVED** [review]
LGTM

**Review by elf-code — APPROVED** [review]
Thanks for the followups!

**Pull request merged**
